### PR TITLE
Clone signature return type nodes if they are present and yield the same type as the signature return type

### PR DESCRIPTION
**Pull request by weswigham**
In #23280 @andy-ms brought up that we never preserve aliases used to create inferred types (which makes sense, given that we immediately collapse type aliases down to their target types). With this PR, we preserve and clone return types of signatures (and thereby their aliases) provided they still yield the same type in the new context. This won't solve the issue in the OP of #23280 (as the alias is annotated within the function body and not as the signature return type node itself), but does make @andy-ms's simplified example work as one may hope.

I'm not 100% sold on this _always_ being a good thing (checkout the type baselines for all the places where `any` became a copy of the erroneous declaration that generated `any`), but it does produce nice results in many places (including preserving input style).

### CODE REVIEW

**Comment by rbuckton** [comment]
@weswigham can you look at the failing test in the build?

File: src/compiler/checker.ts [hunk 1]
```diff
@@ -2797,7 +2797,7 @@ namespace ts {
             }
         }
 
-        function isEntityNameVisible(entityName: EntityNameOrEntityNameExpression, enclosingDeclaration: Node): SymbolVisibilityResult {
+        function resolveNameFromEntityName(entityName: EntityNameOrEntityNameExpression, enclosingDeclaration: Node) {
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Don't we already have a `resolveEntityName` function?

File: src/compiler/checker.ts [hunk 2]
```diff
@@ -3484,6 +3496,39 @@ namespace ts {
                     returnTypeNode = createKeywordTypeNode(SyntaxKind.AnyKeyword);
                 }
                 return createSignatureDeclaration(kind, typeParameters, parameters, returnTypeNode, typeArguments);
+
+                function getDeepSynthesizedCloneAndPaintVisibility<T extends Node>(node: T | undefined): T | undefined {
+                    if (isEntityName(node) || isEntityNameExpression(node)) {
+                        // Paint visible aliases required for copied nodes (so required imports are not elided)
+                        isEntityNameVisible(node, context.enclosingDeclaration);
```
Comment (rbuckton) [inline src/compiler/checker.ts]: Calling `isEntityNameVisible` without using the result is odd. I understand what this is doing, but perhaps its better to split `isEntityNameVisible` into two functions: one that tracks visibility and one that uses the result.

File: src/compiler/checker.ts (hunk 1 above)
Comment (weswigham) [reply src/compiler/checker.ts]: This skips a lot of the work `resolveEntityName` does to verify namespaceyness and also generates an appropriate meaning from the `entityName`'s parent. So they're a bit different, unfortunately.

File: src/compiler/checker.ts (hunk 2 above)
Comment (weswigham) [reply src/compiler/checker.ts]: That'd just be `hasVisibleDeclarations(resolveNameFromEntityName(entityName, enclosingDeclaration), /*shouldComputeAliasToMakeVisible*/ true)`, I suppose?

File: src/compiler/checker.ts (hunk 1 above)
Comment (rbuckton) [reply src/compiler/checker.ts]: We either should have only one function with options to control behavior (preferred), or should clearly document the difference between the two.

**Pull request closed without merging**